wled00/wled_math.cpp

```diff
@@ -49,6 +55,85 @@ float tan_t(float x) {
   return res;
 }
 
+// 16-bit, integer based Bhaskara I's sine approximation: 16*x*(pi - x) / (5*pi^2 - 4*x*(pi - x))
+// input is 16bit unsigned (0-65535), output is 16bit signed (-32767 to +32767)
+// optimized integer implementation by @dedehai
+int16_t sin16_t(uint16_t theta) {
+  int scale = 1;
+  if (theta > 0x7FFF) {
+    theta = 0xFFFF - theta;
+    scale = -1; // second half of the sine function is negative (pi - 2*pi)
+  }
+  uint32_t precal = theta * (0x7FFF - theta);
+  uint64_t numerator = (uint64_t)precal * (4 * 0x7FFF); // 64bit required
+  int32_t denominator = 1342095361 - precal; // 1342095361 is 5 * 0x7FFF^2 / 4
+  int16_t result = numerator / denominator;
+  return result * scale;
+}
+
+int16_t cos16_t(uint16_t theta) {
+  return sin16_t(theta + 16384); //cos(x) = sin(x+pi/2)
+}
+
+uint8_t sin8_t(uint8_t theta) {
+  int32_t sin16 = sin16_t((uint16_t)theta * 257); // 255 * 257 = 0xFFFF
+  sin16 += 0x7FFF; //shift result to range 0-0xFFFF
+  return sin16 >> 8;
```

Comment:
I think a bit of rounding could improve the result - the plot shows that all errors are to the negative, while ideally they should be distributed evenly around 0.

Maybe add something like
``` c++
uint32_t usin16 = sin16; // make sure we're working unsigned
usin16 += 127;                  // prepare for proper rounding
return usin16 >> 8;           // result rounded to nearest positive integer
```

Reply:
good idea. I already tried that and the problem is: overflow.
the returned values are -0x7FFF to + 0x7FFF, sin8 does use the output range 0-255, so the 16bit int needs to be shifted to uint values by addint 0x7FFF. This makes the min value 0 and the max value 0xFFFF. 0xFFFF >> 8 is 255, add 127 and it overflows to 0. To prevent that, an if statement is required, making it slower. Unless there is a clever workaround I am not seeing.

Reply:
Good point 👍 yes 8 did not think about overflows.
I see o e workarounds, not sure which variant would be faster
* `if (usin16 < 65408) usin16 += 127;
* `usin16 = max(usin16, usin16 + 127);`

Reply:
@DedeHai can you test this variant? 

```c++
uint8_t sin8_t(uint8_t theta) {
  int32_t sin16 = sin16_t((uint16_t)theta * 257); // 255 * 257 = 0xFFFF
  sin16 += 0x7FFF + 128; //shift result to range 0-0xFFFF, +128 for rounding
  return min(sin16, int32_t(0xFFFF)) >> 8; // min performs saturation, and prevents overflow
}
```

 It just adds one or two CPU instructions for "min", and should keep the error below +-0.5

Reply:
@DedeHai ok if I push my sin8_t to your branch?